### PR TITLE
[CDAP-20749] Integrate GcpMetadataHandlerInternal with CredentialProviderService

### PR DESCRIPTION
**Pull request by itsankit-google**
JIRA: [CDAP-20749](https://cdap.atlassian.net/browse/CDAP-20749)

This PR does the following:

- Integrates `GCPMetadataHttpHandlerInternal` with `CredentialProviderService`.
- Removes unused `InstantEpochSecondsTypeAdapter`.
- Correctly sets `namespace` in requests from `ConnectionHandler` and `ValidationHandler`.
- Add support for `scopes` in `CredentialProvider` while provisioning credential.
- Adds `ArtifactLocalizerInternalAuthenticator` which sets internal authentication headers using `GcpMetadataTaskContext` as the source.
- Adds `RemoteCredentialProviderService` which is a remote implementation for `CredentialProviderService`.
- Changes `/provision` API from `POST` to `GET`.

[CDAP-20749]: https://cdap.atlassian.net/browse/CDAP-20749?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DistributedPreviewManager.java [hunk 1]
```diff
@@ -217,7 +216,7 @@ public void run() {
             String localhost = InetAddress.getLoopbackAddress().getHostName();
             twillPreparer = twillPreparer.withEnv(PreviewRunnerTwillRunnable.class.getSimpleName(),
                 ImmutableMap.of(
-                    ArtifactLocalizer.GCE_METADATA_HOST_ENV_VAR,
+                    Constants.Preview.GCE_METADATA_HOST_ENV_VAR,
```
Comment (dli357) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DistributedPreviewManager.java:219]: nit: Any particular reason we duplicated this environment variable value?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizerInternalAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import io.cdap.cdap.proto.security.Credential;
+import io.cdap.cdap.proto.security.GcpMetadataTaskContext;
+import java.util.function.BiConsumer;
+import javax.annotation.Nullable;
+
+/**
+ * A class which sets internal authenticated headers for the remote client using a
+ * {@link GcpMetadataTaskContext} as the source.
+ */
+public class ArtifactLocalizerInternalAuthenticator implements InternalAuthenticator {
```
Comment (dli357) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizerInternalAuthenticator.java]: nit: Prefer to rename this to `GcpWorkloadIdentityInternalAuthenticator` to reflect what it does, wdyt?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GcpMetadataHttpHandlerInternal.java [hunk 3]
```diff
@@ -123,29 +138,77 @@ public void token(HttpRequest request, HttpResponder responder,
               METADATA_FLAVOR_HEADER_KEY, METADATA_FLAVOR_HEADER_VALUE));
     }
 
-    // TODO: CDAP-20750
-    if (metadataServiceEndpoint == null) {
+    if (gcpMetadataTaskContext != null) {
```
Comment (dli357) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GcpMetadataHttpHandlerInternal.java]: Should we consider throwing an error here if the metadata context is null? That likely indicates we didn't set something properly.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GcpMetadataHttpHandlerInternal.java [hunk 4]
```diff
@@ -123,29 +138,77 @@ public void token(HttpRequest request, HttpResponder responder,
...
+          fetchTokenFromMetadataServer(scopes).getResponseBodyAsString());
     } catch (Exception ex) {
       LOG.warn("Failed to fetch token from metadata service", ex);
-      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, exceptionToJson(ex),
-          new DefaultHttpHeaders().set(HttpHeaders.CONTENT_TYPE, "application/json"));
+      responder.sendJson(HttpResponseStatus.INTERNAL_SERVER_ERROR, exceptionToJson(ex));
+    }
+  }
+
+  private GcpTokenResponse fetchTokenFromCredentialProvider(String scopes) throws NotFoundException,
```
Comment (dli357) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GcpMetadataHttpHandlerInternal.java:187]: Can we move this to a remote implementation of the `CredentialProvider` class and inject that implementation in the sidecar?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/handler/CredentialProviderHttpHandlerInternal.java [hunk 5]
```diff
@@ -45,8 +47,8 @@
 @Path(Constants.Gateway.INTERNAL_API_VERSION_3)
 public class CredentialProviderHttpHandlerInternal extends AbstractHttpHandler {
 
-  private static final Gson GSON = new GsonBuilder().registerTypeAdapter(Instant.class,
```
Comment (dli357) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/handler/CredentialProviderHttpHandlerInternal.java:48]: The returned ProvisionedCredential still has an instance of `Instant`, why do we want to remove this?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DistributedPreviewManager.java (hunk 1 above)
Comment (itsankit-google) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DistributedPreviewManager.java:219]: This env variable is set in `taskworkertwillrunnable` and `previewrunnertwillrunnable` and not `artifactlocalizertwillrunnable` so wanted to reflect what it does.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizerInternalAuthenticator.java (hunk 2 above)
Comment (itsankit-google) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizerInternalAuthenticator.java]: ack

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/handler/CredentialProviderHttpHandlerInternal.java (hunk 5 above)
Comment (itsankit-google) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/handler/CredentialProviderHttpHandlerInternal.java:48]: Do you see any harm in returning an `Instant` here?

The reason I used it as `Instant`, I am able to easily deserialize the response easily here: https://github.com/cdapio/cdap/pull/15307/files#diff-d98441b91561fbd19d975fbacf290f18c6a18dfc7d7e38d2a9745894b851a3f2R200

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GcpMetadataHttpHandlerInternal.java (hunk 4 above)
Comment (itsankit-google) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GcpMetadataHttpHandlerInternal.java:187]: done.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GcpMetadataHttpHandlerInternal.java (hunk 3 above)
Comment (itsankit-google) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GcpMetadataHttpHandlerInternal.java]: Documented the edge case when it is required:
```
while initializing io.cdap.cdap.common.guice.DFSLocationModule$LocationFactoryProvider#get
in io.cdap.cdap.internal.app.worker.TaskWorkerTwillRunnable.
```

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/handler/CredentialProviderHttpHandlerInternal.java (hunk 5 above)
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/handler/CredentialProviderHttpHandlerInternal.java:48]: SG, didn't realize that Instant was already serializable/deserializable by GSON. Thanks!

File: cdap-master/pom.xml [hunk 6]
```diff
@@ -377,6 +377,12 @@
           <version>${project.version}</version>
           <scope>provided</scope>
         </dependency>
+        <dependency>
```
Comment (dli357) [inline cdap-master/pom.xml:380]: Why does cdap-master need to depend on the extension module?

**Review by dli357 — APPROVED** [review]

File: cdap-master/pom.xml (hunk 6 above)
Comment (dli357) [reply cdap-master/pom.xml:380]: Discussed offline, it's to ensure the extension module is built before cdap-master so it can be packaged properly.

**Pull request merged**